### ISSUE TITLE
Add a library component to the project

### ISSUE THREAD

**Issue by shonfeder** [NathanReb/ocaml-api-watch#12]
Currently the starter code is just in an executable defined in `bin/api_diff.ml`, but we would like to move the core logic into a library, and just leave the outer level of input/output logic in `bin/abi_diff.ml`. 

Moving the core logic into a library will have have at least two benefits:

- it will enable writing unit tests
- provide a library to manipulate the diffs enabling integration with other tools someday

Before we can move the core logic into a library, we need to define a library for the build system. That is the focus of this task.

Steps:

- [ ] Create a new directory `lib` in the root of the repo.
- [ ] Add a file `lib/dune` with a `dune` "stanza" described below.
- [ ] Add a file `lib/api_watch_diff.ml` containing just the dummy function `let diff_interface ~reference ~current = false`.
- [ ] Add the library as a dependency to the executable by editing `bin/dune` to `api_watch_diff` to the the `libraries` field. I.e., the last field will end up as `(libraries api_diff cmdliner compiler-libs.common)`.
- [ ] Finally, reference the dummy function in `bin/api_diff` by adding a top-level declaration `let _ = Api_watch_diff.diff_interface ~reference:"reference" ~current:"current"`. This is to test that the library is accessible.

The contents of `lib/dune` will be this simple [dune library stanza](https://dune.readthedocs.io/en/stable/dune-files.html#library):

```
(library
 (name api_watch_diff)
 (public_name api-watch.diff)
 (libraries compiler-libs.common))
``` 

If every thing works well, you should be able to run

```
dune test
```

with no errors!

**Comment by Siddhi-agg** [comment]
Hi, I have set up project in my system and would like to begin contributing. Can I be assigned this issue?

**Comment by NathanReb** [comment]
Sure thing!

**Comment by Siddhi-agg** [comment]
I followed the tasks given in the issue but when I run 'dune test' the following error shows:
![Screenshot 2024-03-08 053316](https://github.com/NathanReb/ocaml-api-watch/assets/101979598/70db2399-cf34-4a11-a73a-af5a5fea2771)
The 'api_diff' package is not present in the repo. So do I add that package to the repo?

**Comment by shonfeder** [comment]
Hi, @Siddhi-agg! Thats my mistake! I was basing the issue off of the names in `bin/dune`, but I forgot to check the project name :)

For now, lets use `api_watch` for the `name` and `api-watch` for the `public_name`:

```
 (name        api_watch)
 (public_name api-watch)
```

This should resolve the problem, and we can deal with the naming mismatch in a followup!

**Comment by shonfeder** [comment]
(As a tip: it's more helpful to paste in the error message, using code formatting, rather than sharing a screen shot: this is better for accessibility and makes your code snippet searchable.)

**Comment by mridulsrivastavaa** [comment]
> Hi, @Siddhi-agg! Thats my mistake! I was basing the issue off of the names in `bin/dune`, but I forgot to check the project name :)
> 
> For now, lets use `api_watch` for the `name` and `api-watch` for the `public_name`:
> 
> ```
>  (name        api_watch)
>  (public_name api-watch)
> ```
> 
> This should resolve the problem, and we can deal with the naming mismatch in a followup!

Although I have not been assigned to this problem but still I was giving this a try (for getting a follow up for issue 12) but it shows the below error  (if this is not an appreciable practice please let me know)

File "lib/api_diff.ml", line 1, characters 20-29:
1 | let diff_interface ~reference ~current = false
                        ^^^^^^^^^
Error (warning 27 [unused-var-strict]): unused variable reference.
File "lib/api_diff.ml", line 1, characters 31-38:
1 | let diff_interface ~reference ~current = false
                                   ^^^^^^^
Error (warning 27 [unused-var-strict]): unused variable current.
File "bin/dune", line 4, characters 21-29:
4 |  (libraries api_diff cmdliner compiler-libs.common))
                         ^^^^^^^^
Error: Library "cmdliner" not found.



How should I proceed further ?

**Comment by Siddhi-agg** [comment]
Hi @shonfeder changing the names solved the earlier issue.
Now as @mridulsrivastavaa said, it shows the following errors:

root@Ubuntu:/home/vboxuser/ocaml-api-watch# dune test
File "lib/dune", line 4, characters 30-50:
4 |  (libraries api_diff cmdliner compiler-libs.commom))
                                  ^^^^^^^^^^^^^^^^^^^^
Error: Library "compiler-libs.commom" not found.
-> required by library "api-watch" in _build/default/lib
-> required by _build/default/META.api-watch
-> required by alias .api-watch-files
-> required by alias tests/api-diff/runtest
File "bin/api_diff.ml", line 1, characters 8-31:
1 | let _ = Api_diff.diff_interface ~reference:"reference" ~current:"current"
            ^^^^^^^^^^^^^^^^^^^^^^^
Error: Unbound module Api_diff

I tried searching for compiler-libs.common and found it but the name was opam-compiler-libs.common and when I tried to install it using opam, it gave the following error:

[ERROR] Package ocaml-compiler-libs has no version common.

I am also trying to study on the dune stanzas to understand why the Api_diff module is unbound here. Can you point me in the right direction?


> (As a tip: it's more helpful to paste in the error message, using code formatting, rather than sharing a screen shot: this is better for accessibility and makes your code snippet searchable.)

Thanks for the tip! :)
I hope it is better now.

**Comment by NathanReb** [comment]
`api-watch` is the name of the opam package here. The repo is meant to host a variety of tools and libraries.

I edited the issue description to reflect that. Let's use:
```
 (name api_watch_diff)
 (public_name api-watch.diff)
```

In dune you have to prefix the public library name with the package name and a `.` if its name is not the package name. That's why you got this error. Sorry, the issue was a bit misleading here.

Note that `ocaml-compiler-libs` is a repackaging of the compiler libraries. Your problem here is just a typo:
 ```
 (libraries api_diff cmdliner compiler-libs.commom)
 ```
 when it should be:
 ```
 (libraries api_diff cmdliner compiler-libs.common)
 ```

**Comment by mridulsrivastavaa** [comment]
@NathanReb @shonfeder 
My **cmdliner not found** got resolved after I installed the needed. But I still am facing the following error.

File "lib/api_diff.ml", line 1, characters 20-29:
1 | let diff_interface ~reference ~current =false
                        ^^^^^^^^^
Error (warning 27 [unused-var-strict]): unused variable reference.
File "lib/api_diff.ml", line 1, characters 31-38:
1 | let diff_interface ~reference ~current =false
                                   ^^^^^^^
Error (warning 27 [unused-var-strict]): unused variable current.

It says that the variables reference and current are unused. Is there any specific syntax for referencing the dummy function apart from that mentioned in the instructions above ? I couldn't find anything by searching.

**Comment by NathanReb** [comment]
Since this is a dummy function at the moment and we don't use any of the arguments you get a warning from the compiler.

Since we don't want to use them yet you can explicitly tell the compiler you are ignoring them like this:
```ocaml
let diff_interface ~reference:_ ~current:_ = false
```

Also note that you can open a draft pull request, it makes it easier to help you when we can actually see the code.

Finally, when you want to copy paste from your terminal into github, you should do so in a code block. That's a special formatting for code or anything where character alignment matters. It will make them easier to read.

To do that surround the text you want to format that way with triple backquotes ` ``` ` lines. I.e. to get:
```
File "lib/api_diff.ml", line 1, characters 20-29:
1 | let diff_interface ~reference ~current =false
^^^^^^^^^
Error (warning 27 [unused-var-strict]): unused variable reference.
File "lib/api_diff.ml", line 1, characters 31-38:
1 | let diff_interface ~reference ~current =false
^^^^^^^
Error (warning 27 [unused-var-strict]): unused variable current.
```

Write the following:
````md
Here's the error I get when running the command:
```
File "lib/api_diff.ml", line 1, characters 20-29:
1 | let diff_interface ~reference ~current =false
^^^^^^^^^
Error (warning 27 [unused-var-strict]): unused variable reference.
File "lib/api_diff.ml", line 1, characters 31-38:
1 | let diff_interface ~reference ~current =false
^^^^^^^
Error (warning 27 [unused-var-strict]): unused variable current.
```
````

**Comment by NathanReb** [comment]
You can read more about github flavored markdown [here](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax).

**Comment by Siddhi-agg** [comment]
> To clarify: `compiler-libs.common` is a library shipped with the compiler so it is already installed in your switch. You don't need `ocaml-compiler-libs`.

Thanks for the explanation. I understood the difference between the two. 
Sorry, I totally missed the typo while reviewing. I have corrected and submitted a PR for review. I still have some errors I am working on:


```
File "lib/dune", line 4, characters 12-20:
4 |  (libraries api_diff cmdliner compiler-libs.common))
                ^^^^^^^^
Error: Library "api_diff" not found.
-> required by library "api-watch.diff" in _build/default/lib
-> required by _build/default/META.api-watch
-> required by alias .api-watch-files
-> required by alias tests/api-diff/runtest
File "bin/api_diff.ml", line 1, characters 8-37:
1 | let _ = Api_watch_diff.diff_interface ~reference:"reference" ~current:"current"
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Error: Unbound module Api_watch_diff
```

**Comment by MungaSoftwiz** [comment]
I encountered such errors in tests I can say before the `Error: Unbound module API_watch_diff` means that the module Api_watch_diff is not found or not properly imported in our project. So since Api_watch_diff is a module defined in our project, ensure it's properly implemented and included in our project files. You can check the file where Api_watch_diff is defined in the `lib` directory and verify that it's correctly imported or included.

I can see from your draft PR that the Api_watch_diff is a file but it's in small caps so maybe using a capatilased `A` might be causing the issue. And also checking the dune file I notice the module starts with a small `a`. 

So something with the proper module name would solve this issue. Check below:
```
let _ = api_watch_diff.diff_interface ~reference:"reference" ~current:"current"
```

**Comment by mridulsrivastavaa** [comment]
> Since this is a dummy function at the moment and we don't use any of the arguments you get a warning from the compiler.
> 
> Since we don't want to use them yet you can explicitly tell the compiler you are ignoring them like this:
> 
> ```ocaml
> let diff_interface ~reference:_ ~current:_ = false
> ```
> 
> Also note that you can open a draft pull request, it makes it easier to help you when we can actually see the code.
> 
> Finally, when you want to copy paste from your terminal into github, you should do so in a code block. That's a special formatting for code or anything where character alignment matters. It will make them easier to read.
> 
> To do that surround the text you want to format that way with triple backquotes `` ``` `` lines. I.e. to get:
> 
> ```
> File "lib/api_diff.ml", line 1, characters 20-29:
> 1 | let diff_interface ~reference ~current =false
> ^^^^^^^^^
> Error (warning 27 [unused-var-strict]): unused variable reference.
> File "lib/api_diff.ml", line 1, characters 31-38:
> 1 | let diff_interface ~reference ~current =false
> ^^^^^^^
> Error (warning 27 [unused-var-strict]): unused variable current.
> ```
> 
> Write the following:
> 
> ```
> Here's the error I get when running the command:
> ```
> File "lib/api_diff.ml", line 1, characters 20-29:
> 1 | let diff_interface ~reference ~current =false
> ^^^^^^^^^
> Error (warning 27 [unused-var-strict]): unused variable reference.
> File "lib/api_diff.ml", line 1, characters 31-38:
> 1 | let diff_interface ~reference ~current =false
> ^^^^^^^
> Error (warning 27 [unused-var-strict]): unused variable current.
> ```
> ```

Thank you for the code formatting tip. I'll make sure of using this from the next time.

**Comment by Siddhi-agg** [comment]
> I can see from your draft PR that the Api_watch_diff is a file but it's in small caps so maybe using a capatilased `A` might be causing the issue. And also checking the dune file I notice the module starts with a small `a`.

Hey @MungaSoftwiz! I tried your suggestion, but the issue was due to the fact that api_diff.ml was renamed to api_watch_diff.ml, but in the dune file, the libraries dependency didn't have the name changed. So when I changed it, the issue was resolved.

Thanks for helping tho!

**Comment by shonfeder** [comment]
Closed in #20
